mmdet/models/roi_heads/sparse_roi_head.py

```diff
@@ -0,0 +1,338 @@
+                gt_labels,
+                self.train_cfg[stage],
+                True,
+            )
+            cls_score = bbox_results['cls_score']
+            decode_bbox_pred = bbox_results['decode_bbox_pred']
+
+            single_stage_loss = self.bbox_head[stage].loss(
+                cls_score.view(-1, cls_score.size(-1)),
+                decode_bbox_pred.view(-1, 4),
+                *bbox_targets,
+                imgs_whwh=imgs_whwh)
+            for key, value in single_stage_loss.items():
+                all_stage_loss[f'stage{stage}_{key}'] = value * \
+                                    self.stage_loss_weights[stage]
+            object_feats = bbox_results['object_feats']
+
+        return all_stage_loss
+
+    def simple_test(self,
+                    x,
+                    proposal_boxes,
+                    proposal_feats,
+                    img_metas,
+                    imgs_whwh,
+                    rescale=False):
+        """Test without augmentation.
+
+        Args:
+            x (list[Tensor]): list of multi-level img features.
+            proposals (Tensor): Decoded proposal bboxes, has shape
+                (batch_size, num_proposals, 4)
+            proposal_features (Tensor): Expanded proposal
+                features, has shape
+                (batch_size, num_proposals, proposal_feature_channel)
+            img_metas (dict): meta information of images.
+            imgs_whwh (Tensor): Tensor with shape (batch_size, 4),
+                    the dimension means
+                    [img_width,img_height, img_width, img_height].
+            rescale (bool): If True, return boxes in original image
+                space. Defaults to False.
+
+        Returns:
+            bbox_results (list[tuple[np.ndarray]]): \
+                [[cls1_det, cls2_det, ...], ...]. \
+                The outer list indicates images, and the inner \
+                list indicates per-class detected bboxes. The \
+                np.ndarray has shape (num_det, 5) and the last \
+                dimension 5 represents (x1, y1, x2, y2, score).
+        """
+        assert self.with_bbox, 'Bbox head must be implemented.'
+        # Decode initial proposals
+        num_imgs = len(img_metas)
+        num_proposals = proposal_boxes.size(1)
+        proposal_list = [proposal_boxes[i] for i in range(num_imgs)]
+        object_feats = proposal_feats
+        for stage in range(self.num_stages):
+            rois = bbox2roi(proposal_list)
+            bbox_results = self._bbox_forward(stage, x, rois, object_feats,
+                                              img_metas)
+            object_feats = bbox_results['object_feats']
+            cls_score = bbox_results['cls_score']
+            proposal_list = bbox_results['detach_proposal_list']
+
+        num_classes = self.bbox_head[-1].num_classes
+        det_bboxes = []
+        det_labels = []
+
+        if self.bbox_head[-1].loss_cls.use_sigmoid:
```

Comment:
We may refactor lines 270 - 305 since there are some duplicated codes. The only different part seems to be in the generation of indices.